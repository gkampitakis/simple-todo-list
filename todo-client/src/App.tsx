import React, { useEffect, useState } from 'react';
import { createStyles, Grid, makeStyles } from '@material-ui/core';
import { Header, InputForm, List } from './components';
import { Todo } from './types';

const useStyles = makeStyles(() =>
  createStyles({
    root: {
      height: '90vh',
      flexGrow: 1
    },
    container: {
      alignItems: 'center',
      height: '100%'
    }
  })
);

function App() {
  const [items, setItems] = useState<Todo[]>([]);
  const classes = useStyles();

  useEffect(() => {
    fetch('http://localhost:4000/todo/all')
      .then(async (response) => {
        const data = await response.json();

        setItems(data);
      })
      .catch(console.error);
  }, []);

  function removeItem(id: string) {
    setItems((items) => items.filter((item) => item.id !== id));
  }

  return (
    <>
      <Header />
      <div className={classes.root}>
        <Grid container direction={'column'} className={classes.container}>
          <InputForm insert={setItems} />
          <List list={items} removeItem={removeItem} />
        </Grid>
      </div>
    </>
  );
}

export default App;
