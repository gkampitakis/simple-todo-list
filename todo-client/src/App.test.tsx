import React from 'react';
import {
  render,
  screen,
  waitForElementToBeRemoved
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
// import { FetchMock } from 'jest-fetch-mock/types';
import { fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import App from './App';

// const fetchMock = fetch as FetchMock;

const mockData = [
  {
    id: '1',
    todo: 'hello world',
    createdAt: '1234'
  },
  {
    id: '2',
    todo: 'hello world 2',
    createdAt: '1234'
  },
  {
    id: '3',
    todo: 'hello world 3',
    createdAt: '1234'
  }
];

describe('Todo list', () => {
  beforeAll(fetchMock.enableMocks);
  beforeEach(() => {
    fetchMock.resetMocks();
  });

  it('Should render consistently', async () => {
    fetchMock.mockResponseOnce(JSON.stringify(mockData));
    await act(async () => {
      const { baseElement } = render(<App />);

      await screen.findByText('hello world');

      expect(baseElement).toMatchSnapshot();
    });
  });

  it('Should log in case of error when fetching', async () => {
    expect.assertions(1);
    const error = Error('mock message');
    fetchMock.mockReject(error);
    const errorSpy = jest.spyOn(console, 'error');
    await act(async () => {
      render(<App />);

      setImmediate(() => {
        expect(errorSpy).toHaveBeenCalledWith(error);
      });
    });
  });

  it('Should delete item from list', async () => {
    fetchMock.mockResponseOnce(JSON.stringify(mockData));
    fetchMock.mockResponse('', { status: 204 });
    await act(async () => {
      const { baseElement } = render(<App />);

      await screen.findByText('hello world');

      userEvent.click(screen.getByTestId('delete_2'));
      userEvent.click(screen.getByTestId('delete_3'));

      await waitForElementToBeRemoved(screen.getByTestId('delete_3'));

      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('Input Form', () => {
    it('Should create a new entry', async () => {
      fetchMock.mockResponseOnce(JSON.stringify(mockData));
      fetchMock.mockResponseOnce(
        JSON.stringify({
          id: 4,
          todo: 'new item',
          createdAt: '123231'
        })
      );

      await act(async () => {
        render(<App />);

        await screen.findByText('hello world');

        userEvent.type(screen.getByPlaceholderText(/insert a /i), 'new item');
        userEvent.click(screen.getByTestId('save-btn'));

        await screen.findByText('new item');

        expect(fetchMock).toHaveBeenCalledWith('http://localhost:4000/todo', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            todo: 'new item'
          })
        });
      });
    });

    it('Should create a new entry with enter', async () => {
      fetchMock.mockResponseOnce(JSON.stringify(mockData));
      fetchMock.mockResponseOnce(
        JSON.stringify({
          id: 4,
          todo: 'new item',
          createdAt: '123231'
        })
      );

      await act(async () => {
        render(<App />);

        await screen.findByText('hello world');

        userEvent.type(screen.getByPlaceholderText(/insert a /i), 'new item');
        fireEvent.keyDown(screen.getByPlaceholderText(/insert a /i), {
          key: 'Enter'
        });

        await screen.findByText('new item');

        expect(fetchMock).toHaveBeenCalledWith('http://localhost:4000/todo', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            todo: 'new item'
          })
        });
      });
    });
  });
});
