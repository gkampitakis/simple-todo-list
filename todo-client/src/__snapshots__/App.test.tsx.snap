// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo list Should delete item from list 1`] = `
<body>
  <div>
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <h6
          class="MuiTypography-root makeStyles-title-17 MuiTypography-h6"
        >
          Simple Todo List üìù
        </h6>
      </div>
    </header>
    <div
      class="makeStyles-root-15"
    >
      <div
        class="MuiGrid-root makeStyles-container-16 MuiGrid-container MuiGrid-direction-xs-column"
      >
        <div
          class="MuiPaper-root MuiCard-root makeStyles-inputForm-18 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiFormControl-root MuiTextField-root makeStyles-inputField-20"
          >
            <div
              class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiFilledInput-input"
                placeholder="Insert a todo item ..."
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-saveBtn-19 MuiButton-containedPrimary Mui-disabled Mui-disabled"
            data-testid="save-btn"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Save
            </span>
          </button>
        </div>
        <ul
          class="MuiList-root makeStyles-root-21 MuiList-padding"
        >
          <div>
            <li
              class="MuiListItem-container"
            >
              <div
                class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
              >
                <div
                  class="MuiListItemAvatar-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    hello world
                  </span>
                </div>
              </div>
              <div
                class="MuiListItemSecondaryAction-root"
              >
                <button
                  aria-label="delete"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
                  data-testid="delete_1"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorError"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </li>
            <hr
              class="MuiDivider-root"
            />
          </div>
        </ul>
      </div>
    </div>
  </div>
</body>
`;

exports[`Todo list Should render consistently 1`] = `
<body>
  <div>
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <h6
          class="MuiTypography-root makeStyles-title-3 MuiTypography-h6"
        >
          Simple Todo List üìù
        </h6>
      </div>
    </header>
    <div
      class="makeStyles-root-1"
    >
      <div
        class="MuiGrid-root makeStyles-container-2 MuiGrid-container MuiGrid-direction-xs-column"
      >
        <div
          class="MuiPaper-root MuiCard-root makeStyles-inputForm-4 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiFormControl-root MuiTextField-root makeStyles-inputField-6"
          >
            <div
              class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiFilledInput-input"
                placeholder="Insert a todo item ..."
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-saveBtn-5 MuiButton-containedPrimary Mui-disabled Mui-disabled"
            data-testid="save-btn"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Save
            </span>
          </button>
        </div>
        <ul
          class="MuiList-root makeStyles-root-7 MuiList-padding"
        >
          <div>
            <li
              class="MuiListItem-container"
            >
              <div
                class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
              >
                <div
                  class="MuiListItemAvatar-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    hello world
                  </span>
                </div>
              </div>
              <div
                class="MuiListItemSecondaryAction-root"
              >
                <button
                  aria-label="delete"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
                  data-testid="delete_1"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorError"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </li>
            <hr
              class="MuiDivider-root"
            />
          </div>
          <div>
            <li
              class="MuiListItem-container"
            >
              <div
                class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
              >
                <div
                  class="MuiListItemAvatar-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    hello world 2
                  </span>
                </div>
              </div>
              <div
                class="MuiListItemSecondaryAction-root"
              >
                <button
                  aria-label="delete"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
                  data-testid="delete_2"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorError"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </li>
            <hr
              class="MuiDivider-root"
            />
          </div>
          <div>
            <li
              class="MuiListItem-container"
            >
              <div
                class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
              >
                <div
                  class="MuiListItemAvatar-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    hello world 3
                  </span>
                </div>
              </div>
              <div
                class="MuiListItemSecondaryAction-root"
              >
                <button
                  aria-label="delete"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
                  data-testid="delete_3"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorError"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </li>
            <hr
              class="MuiDivider-root"
            />
          </div>
        </ul>
      </div>
    </div>
  </div>
</body>
`;
